grpc/grpc-interceptor.go

```diff
@@ -0,0 +1,110 @@
+package grpc
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"runtime"
+	"strings"
+
+	"google.golang.org/grpc"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+)
+
+//
+// Digma's grpc-interceptor:
+// should be chained after the original otelgrpc interceptors.
+//
+// usage for example :
+//  import (
+//     	"google.golang.org/grpc"
+//      "go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc"
+//  	digma_grpc "github.com/digma-ai/otel-go-instrumentation/grpc"
+//   	grpc_middleware "github.com/grpc-ecosystem/go-grpc-middleware"
+//  )
+//
+//	server := grpc.NewServer(
+//	    grpc.UnaryInterceptor(grpc_middleware.ChainUnaryServer(
+//	       otelgrpc.UnaryServerInterceptor(),
+//	       digmagrpc.UnaryServerInterceptor(),
+//      )),
+//      grpc.StreamInterceptor(grpc_middleware.ChainStreamServer(
+//	      otelgrpc.StreamServerInterceptor(),
+//	      digma_grpc.StreamServerInterceptor(),
+//      )),
+//  )
+//
+
+
+// UnaryServerInterceptor returns a grpc.UnaryServerInterceptor suitable
+// for use in a grpc.NewServer call.
+func UnaryServerInterceptor() grpc.UnaryServerInterceptor {
+	return func(
+		ctx context.Context,
+		req interface{},
+		info *grpc.UnaryServerInfo,
+		handler grpc.UnaryHandler,
+	) (interface{}, error) {
+
+		methodFqn := buildMethodFqn(info.Server, info.FullMethod)
+		//TODO: debug line, remove it
+		fmt.Println("methodFqn:", methodFqn)
+
+		span := trace.SpanFromContext(ctx)
+		span.SetAttributes(attribute.String("endpoint.function_full_name", methodFqn))
+
+		// standard call of interceptor
+		resp, err := handler(ctx, req)
+		return resp, err
+	}
+}
+
+// StreamServerInterceptor returns a grpc.StreamServerInterceptor suitable
+// for use in a grpc.NewServer call.
+func StreamServerInterceptor() grpc.StreamServerInterceptor {
+	return func(
+		srv interface{},
+		ss grpc.ServerStream,
+		info *grpc.StreamServerInfo,
+		handler grpc.StreamHandler,
+	) error {
+
+		methodFqn := buildMethodFqn(srv, info.FullMethod)
+		//TODO: debug line, remove it
+		fmt.Println("methodFqn:", methodFqn)
+
+		ctx := ss.Context()
+		span := trace.SpanFromContext(ctx)
+		span.SetAttributes(attribute.String("endpoint.function_full_name", methodFqn))
+
+		// standard call of interceptor
+		err := handler(srv, ss)
+		return err
+	}
+}
+
+func methodOnly(fullMethod string) string {
+	ix := strings.LastIndex(fullMethod, "/")
+
+	return fullMethod[ix+1:]
+}
+
+func fqnOfServiceWithMethod(srv interface{}, methodName string) string {
+	typeOfService := reflect.TypeOf(srv)
+	methodRef, ok := typeOfService.MethodByName(methodName)
+	if !ok {
+		panic("cannot find method name '" + methodName + "' for service '" + typeOfService.Name() + "'")
```

Comment:
can we return the error all the way from the interceptor and not raise panic?

Reply:
I think that no one can handle the error of the interceptor, so raising it up won't help.
I will log it as error instead of panic - so we would do less harm to our customers => **runtime error indeed is bad**